sdk_refactor_design_proposal.md

```diff
@@ -0,0 +1,392 @@
+
+For these reasons, I propose we extract the configuration and credential information to their own objects to be created by the user before the Client is called. In those classes, all the collection and validations will be performed. As a result, the Client constructor will not be forced to be in charge of handle multiple responsibilities, the user will have to explicitly define the and the areas of volatility will be contained.
+
+We can handle the separation in one of the following ways:
+
+|      | Solution                                        | Description                                                  | Pros                                                         | Cons                                                         |
+| ---- | ----------------------------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| 1    | Keep as is                                      |                                                              | - Doesn't break current user apps<br />- Seems to be the direction the industry is going | - 92 lines and 9 params so far makes it unweidy<br />- Constructor is overloaded, handling multiple responsibilities<br />- Doesn't take advantage of OOP |
+| 2    | Config/Creds objects for *each* supported flow  | ConfigurationFromFile()  - .conjurrc<br />ConfigurationFromEnv() - ENV vars<br />CredentialsFromFile() - netrc<br />CredentialsFromKeystore() - keystore | - Explicit, the flow the user chooses is very clear<br />- Security is up to user's the discretion | - Conflicting experience with CLI. The CLI performs a discovery on Config/Creds retrieval type<br />- Requires changes when a user wants to choose another flow |
+| 3    | Generic Config/Creds object                     | Configuration()<br />Credentials()<br />and we will discover where the connection details live based on environment's context | - Easy maintainability. No change in user's code as we build new ways to fetch config and creds<br />- Aligns with CLI UX of Config/Cred discovery | - Implicit and the flow we choose is not clear               |
+|      | A single Config/Cred object with defining param | Configuration(action=file) - .conjurrc<br />Credential(action=keystore) | - Explicit<br />- Minimal code changes should the user choose a different flow<br />- Strong drive in Python to be as explicit as possible | - Requires changes in code if user wants to adopt another retrieval method<br />- Conflicting experience with CLI. The CLI preforms a discovery on Config/Creds retrieval type |
+
+**Decision:** Option #4. 
+
+The flow for configuration and credential will be determined *explicitly* by enabling the user to pass in the flow of their choosing via a parameter to the object during instantiation. 
```

Comment:
Why would we want to do that? The BL want config (which creds are part of), why does it matter if those are taken from a file or the keystore?

Reply:
For the sake of being as explicit as possible so the user doesn't have to guess which context their environment is running in. 
Do you suggest, performing a discovery is better? For example have the user provide a general `creds = conjur.Credentials(username, password, url), and we search for them through mediums we support (netrc, keystore)?

Reply:
I don't think the sdk should support .netrc or .conjurrc files. it is up for the user to provide this data to the SDK at runtime.
.netrc is not safe and I don't think we should keep supporting it in the sdk.